drivers/docker/reconciler.go

```diff
@@ -0,0 +1,175 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+func (d *Driver) removeDanglingContainersGoroutine() {
+	if !d.config.GC.DanglingContainers.Enabled {
+		d.logger.Debug("skipping dangling containers handling; is disabled")
+		return
+	}
+
+	period := d.config.GC.DanglingContainers.period
+
+	succeeded := true
+
+	// ensure that we wait for at least a period or creation timeout
+	// for first container GC iteration
+	// The initial period is a grace period for restore allocation
+	// before a driver may kill containers launched by an earlier nomad
+	// process.
+	initialDelay := period
+	if d.config.GC.DanglingContainers.creationTimeout > initialDelay {
+		initialDelay = d.config.GC.DanglingContainers.creationTimeout
+	}
+
+	timer := time.NewTimer(initialDelay)
+	for {
+		select {
+		case <-timer.C:
+			if d.previouslyDetected() && d.fingerprintSuccessful() {
+				err := d.removeDanglingContainersIteration()
+				if err != nil && succeeded {
+					d.logger.Warn("failed to remove dangling containers", "error", err)
+				}
+				succeeded = (err == nil)
+			}
+
+			timer.Reset(period)
+		case <-d.ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *Driver) removeDanglingContainersIteration() error {
+	tracked := d.trackedContainers()
+	untracked, err := d.untrackedContainers(tracked, d.config.GC.DanglingContainers.creationTimeout)
+	if err != nil {
+		return fmt.Errorf("failed to find untracked containers: %v", err)
+	}
+
+	for _, id := range untracked {
+		d.logger.Info("removing untracked container", "container_id", id)
+		err := client.RemoveContainer(docker.RemoveContainerOptions{
+			ID:    id,
+			Force: true,
+		})
+		if err != nil {
+			d.logger.Warn("failed to remove untracked container", "container_id", id, "error", err)
+		}
+	}
+
+	return nil
+}
+
+// untrackedContainers returns the ids of containers that suspected
+// to have been started by Nomad but aren't tracked by this driver
+func (d *Driver) untrackedContainers(tracked map[string]bool, creationTimeout time.Duration) ([]string, error) {
+	result := []string{}
+
+	cc, err := client.ListContainers(docker.ListContainersOptions{})
```

Comment:
I believe [only running containers are listed by default](https://docs.docker.com/engine/api/v1.23/#21-containers) which means we won't GC stopped containers.

Although GCing stopped containers introduces 2 other things to consider:

1. We should only GC containers with [`taskHandle.removeContainerOnExit`](https://www.nomadproject.io/docs/drivers/docker.html#container) set.
2. There is a race between this goroutine and task runner stopping and removing the container itself. We may need a stop grace period similar to the create grace period to avoid spurious error logs when this race is hit and the container is removed twice. Although maybe it's not worth the complexity?

Regardless of desired behavior around GCing stopped containers, we should add a test with a stopped container.